src/exchanges/KyberAdapter.sol

```diff
@@ -1,201 +1,189 @@
+    /// @param _orderValues [0] Maker asset quantity
+    /// @param _orderValues [1] Taker asset quantity
+    /// @param _orderValues [6] Taker asset fill quantity
     function takeOrder(
-        address targetExchange,
-        address[8] memory orderAddresses,
-        uint[8] memory orderValues,
-        bytes[4] memory orderData,
-        bytes32 identifier,
-        bytes memory signature
-    ) public override onlyManager notShutDown {
-        Hub hub = getHub();
-
-        require(
-            orderValues[1] == orderValues[6],
-            "fillTakerQuantity must equal takerAssetQuantity"
-        );
-
-        address makerAsset = orderAddresses[2];
-        address takerAsset = orderAddresses[3];
-        uint makerAssetAmount = orderValues[0];
-        uint takerAssetAmount = orderValues[1];
-
-        uint minRate = calcMinRate(
-            takerAsset,
-            makerAsset,
-            takerAssetAmount,
-            makerAssetAmount
-        );
+        address _targetExchange,
+        address[8] memory _orderAddresses,
+        uint[8] memory _orderValues,
+        bytes[4] memory _orderData,
+        bytes32 _identifier,
+        bytes memory _signature
+    )
+        public
+        override
+    {
+        __validateTakeOrderParams(_orderValues);
 
-        uint actualReceiveAmount = dispatchSwap(
-            targetExchange, takerAsset, takerAssetAmount, makerAsset, minRate
-        );
-        require(
-            actualReceiveAmount >= makerAssetAmount,
-            "Received less than expected from Kyber swap"
-        );
+        (
+            address[] memory fillAssets,
+            uint256[] memory fillExpectedAmounts
+        ) = __formatFillTakeOrderArgs(_orderAddresses, _orderValues);
 
-        getAccounting().addAssetToOwnedAssets(makerAsset);
-        getAccounting().updateOwnedAssets();
-        getTrading().returnAssetToVault(makerAsset);
-        getTrading().orderUpdateHook(
-            targetExchange,
-            bytes32(0),
-            Trading.UpdateType.take,
-            [payable(makerAsset), payable(takerAsset)],
-            [actualReceiveAmount, takerAssetAmount, takerAssetAmount]
-        );
+        __fillTakeOrder(_targetExchange, fillAssets, fillExpectedAmounts);
     }
 
     // INTERNAL FUNCTIONS
 
-    /// @notice Call different functions based on type of assets supplied
-    function dispatchSwap(
-        address targetExchange,
-        address srcToken,
-        uint srcAmount,
-        address destToken,
-        uint minRate
+    // Minimum acceptable rate of taker asset per maker asset
+    function __calcMinMakerAssetPerTakerAssetRate(
+        address[] memory _fillAssets,
+        uint256[] memory _fillExpectedAmounts
     )
         internal
-        returns (uint actualReceiveAmount)
+        view
+        returns (uint)
     {
+        return mul(
+            _fillExpectedAmounts[1],
+            10 ** uint256(ERC20WithFields(_fillAssets[1]).decimals())
```

Comment:
Good catch! Yeah, this is what I was porting from: https://github.com/melonproject/protocol/blob/532e549f5b97ea1e95e465bb4204c44d44350154/src/prices/KyberPriceFeed.sol#L293

Should we put our basic price calculation along with other basic calcs that we use frequently in a simple `MelonMath` contract that we can inherit for consistency?

Fixed for now, but leaving this open in case we want to address that

Reply:
it might make sense, and feel free to open an issue for it if you can identify a bunch of different calculations we do of identical form. My only worry is that it might become too abstract when just seeing the plain math will do, so as long as we avoid that.